### DISCUSSION TITLE
What is the ordering of the output of e3nn.o3.spherical_harmonics?

### DISCUSSION THREAD

**Discussion by esslushy** [e3nn/e3nn#419] [Q&A] [answered]
So I am working with some spherical harmonics where the ordering matters. I have the following code:
```
from e3nn.o3 import Irreps, spherical_harmonics
l = Irreps.spherical_harmonics(1)
r_ij = torch.randn(6, 3)
sph_har = spherical_harmonics(l, r_ij, normalize=True, normalization="integral")
```
In the resulting `sph_har`, what is the ordering of the spherical harmonic results. Would it be:
sph_har[:, 0]: l=0, m=0
sph_har[:, 1]: l=1, m=-1
sph_har[:, 2]: l=1, m=0
sph_har[:, 3]: l=1, m=1

Or
sph_har[:, 0]: l=0, m=0
sph_har[:, 1]: l=1, m=0
sph_har[:, 2]: l=1, m=1
sph_har[:, 3]: l=1, m=-1

Or any other variation. I assume it is the first result, but I wanted to double check to make sure. If it is not in the first style, is there a way to make it output to match that style? Or what is the proper output ordering produced by the spherical harmonics function?

**Comment by mariogeiger** [comment, marked as answer]
Yes it's the first option. 

Be careful that the input is different than usual. You need to input something like x z y to recover the usual expressions of the real spherical harmonics you can find on Wikipedia (I'm not sure of the order of x and y and potential sign flip but for sure z is in the center)

**Reply by esslushy** [reply]
Just to clarify. The documentation says I need to put it in the order of y, z, x [[ref](https://docs.e3nn.org/en/stable/api/o3/o3_sh.html)]. In order to match the wikipedia table. Assuming `r_ij` is in [x, y, z] form, I would need to reorder it via `r_ij[:, [1, 2, 0]]` so that it would be input in the form [y, z, x] in order to get the spherical harmonic order expected in the first list?

**Reply by liangzhixin-202169** [reply]
I also think an order of r_ij[:, [1, 2, 0]] is reasonable. However, the result of the "Convolution" section in the manual(https://docs.e3nn.org/en/stable/guide/convolution.html) will be non equivariance if "edge_vec" is changed to "edge_vec[:, [1,2,0]]"

**Comment by liangzhixin-202169** [comment]
The order seems to be different with the first option.
`ir = o3.Irreps(o3.Irrep(1, -1))
sh = o3.spherical_harmonics(ir,
                              torch.tensor(
                                  [[2, 0, 0],
                                   [0, 2, 0],
                                   [0, 0, 2]], dtype=torch.float32),
                              normalize=True,
                              normalization='norm')
print(sh)`
the output is:
`tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]])`
which is the order of [x,y,z], corresponding to Y[1,+1], Y[1,-1], Y[1,0] in the table of wikipedia
